apps/bgu/CMakeLists.txt

```diff
@@ -4,10 +4,13 @@ project(bgu)
 enable_testing()
 
 # Set up language settings
-set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED YES)
 set(CMAKE_CXX_EXTENSIONS NO)
 
+# MSVC won't set __cplusplus correctly unless this flag is also set
```

Comment:
Is there no way to hide this idiocy from our users?

Reply:
We reported this to the CMake folks and they insisted they couldn't change it without a 'policy' flag to enable it, because existing (broken) code might rely on this (insanely broken) behavior, so they aren't gonna change it anytime soon.

The alternative that most people seem to end up using is to combine sniffing of `__cplusplus` with sniffing of `_MSC_LANG`, which would work I guess, but is just so f***ing wrong that I didn't want to do it, just on principle. Maybe that's a more pragmatic approach than requiring the build files to insert extra things to work around bad decisions by both MSVC and CMake.

Reply:
Can we move the setting onto our side of the fence somehow, instead of in the user CMakeLists?

Reply:
Well, yes and no.

We currently don't *need* to check __cplusplus at all (there's a reality check in Codegen_LLVM.cpp but it is just as a backstop).

If we only used C++17 features internal to libHalide, then we wouldn't need this.

If we eventually use a C++17 feature in headers exposed to user code (e.g. in templates), though, then we'd definitely have this problem (ie the user code's build system would have to set this flag).

I guess the better answer is to bite the bullet and accept that checks of that sort will have to check both __cplusplus and _MSC_LANG for the foreseeable future, which is just a sorry state of affairs.

Reply:
(fix pushed)